src/Silk.Core.Discord/EventHandlers/Guilds/GuildEventHandlerService.cs

```diff
@@ -0,0 +1,178 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using DSharpPlus;
+using DSharpPlus.Entities;
+using MediatR;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Silk.Core.Data.MediatR.Guilds;
+using Silk.Core.Data.MediatR.Users;
+using Silk.Core.Data.Models;
+using Silk.Core.Discord.EventHandlers.Notifications;
+using Silk.Core.Discord.Types;
+using Silk.Extensions;
+using Silk.Shared.Constants;
+
+namespace Silk.Core.Discord.EventHandlers.Guilds
+{
+    public class GuildEventHandlerService : IHostedService
+    {
+        public ConcurrentQueue<Lazy<Task>> CacheQueue { get; } = new();
+
+        private DateTime? _startTime;
+
+        private bool _cachedAllInitialGuilds;
+        private bool _logged;
+
+        private readonly IMediator _mediator;
+        private readonly DiscordShardedClient _client;
+        private readonly ILogger<GuildEventHandlerService> _logger;
+
+        private Dictionary<int, int> _guilds;
+
+        private int _shardCount; // How many shards to wait for. //
+        private int _currentShardsCompleted; // How many shards have fired GUILD_DOWNLOAD_COMPLETE. //
+        public GuildEventHandlerService(IMediator mediator, DiscordShardedClient client, ILogger<GuildEventHandlerService> logger)
+        {
+            _mediator = mediator;
+            _client = client;
+            _logger = logger;
+        }
+
+        private const string OnGuildJoinThankYouMessage = "Hiya! My name is Silk! I hope to satisfy your entertainment and moderation needs. I respond to mentions and `s!` by default, but you can change the prefix by using the prefix command.\n" +
```

Comment:
Nit: Should use a verbatim string literal for this instead. Will make it look cleaner.